### ISSUE TITLE
Fuel consumption

### ISSUE THREAD

**Issue by mugling** [CleverRaven/Cataclysm-DDA#15775]
IRL a reasonable **approximation** of fuel consumption would be 
- 1000kg vehicle
- 100km/h
- 60L fuel
- 600km

```
k = 0.1ml/Kg/km
fuel = k * weight * distance
```

Taking an (outdoor) cataclysm tile as ~100m and presuming the `safe speed` is the most efficient:

```
k = 0.01ml/Kg/tile
```

`k` represents engine efficiency and should not differ by more than order of magnitude from the best to worst engine. When rescaled as an integral value:

| engine | k x10⁻³ |
| --- | --- |
| 1-cylinder engine | 3 |
| V-twin engine | 9 |
| Inline-4 engine | 10 |
| V6 engine | 8 |
| V8 engine | 6 |
| V12 engine | 2 |

Adding more displacement to an engine increases `safe speed` but doesn't affect efficiency.

Traveling outside this ideal speed adversely affects fuel efficiency. A normal distribution models this reasonably well (IRL this would be biased against higher speeds).

![normal_distribution](https://cloud.githubusercontent.com/assets/12238606/13730285/1d7b11ce-e943-11e5-8415-9cde912b261b.png)

A linear implementation is almost as good and much simpler, taking `v` to be velocity, `s` to be safe speed and `q` to be the proportional efficiency

```
q = 1 - ( |( v - s )| / s )
```

As c++ code

```
double q = 1.0 - ( double( std::abs( v - s ) )  / s )
```

Now lets ignore going under the `safe speed` as most of our vehicles aren't controllable at that speed (and that's a separate issue). We can use the `vehicle::strain()` function...

```
double vehicle::fuel_consumption( double k, int distance ) {
    return k * ( strain() + 1.0 ) * distance * total_mass();
}
```

Not a perfect system (ignores acceleration, terrain and k-mass) however it does result in sensible fuel consumption and adds the concept of power:weight ratio for engines.

**tldr; a 1000kg car with a Inline-4 engine and a full tank of fuel can travel a distance of 6000 tiles when driven at the safe speed**

**Comment by Coolthulhu** [comment]
> Adding more displacement to an engine increases safe speed but doesn't affect efficiency.

I don't like that part. Displacement doesn't have any downsides. Having it boost speed without any negative effects would keep the existing "bigger is better" thing when it comes to displacement.

The more I think about the variable bigness vehicle part system the less I like it - both for wheels and for engines.

Linear sounds too painful at low velocity and could lead to starting consuming more fuel than the travel itself. Something close-to-but-not-linear could be good, though maybe instead of that (or in addition to that?), big engines could have significantly larger fuel consumption on idling.

Safe speed shouldn't be the optimal, it is the maximum speed at which the engine doesn't take damage. Optimal speed should probably be much lower.

Not sure about linearly scaling fuel consumption with mass. IMO it should only depend on how loaded the engine is multiplied by some measure of engine size (multiplied by efficiency factor), a tiny engine pulling a giant truck incredibly slowly shouldn't be able to drain the entire tank in seconds.

**Comment by mugling** [comment]
> Displacement doesn't have any downsides

This is problematic. The larger displacement engines should at least weigh more and be louder.

> Something close-to-but-not-linear could be good

I was struggling with the math. Any suggestions?

> Optimal speed should probably be much lower.

Chosen as it exists at present. Again open to reasonable suggestions. Possibly half safe speed?

> a tiny engine pulling a giant truck incredibly slowly shouldn't be able to drain the entire tank in seconds.

I don't think it can as the algorithm depends on distance and such a vehicle should have incredibly poor (or no velocity) so as to not actually travel any distance

**Comment by Coolthulhu** [comment]
> The larger displacement engines should at least weigh more and be louder.

Well OK, it is slightly louder, as noise is currently calculated off maximum power, but the difference isn't big.

> I was struggling with the math. Any suggestions?

Few ideas:
* `consumption = max_consumption * (percent of engine power used + constant)`
* `consumption = max_consumption * (sqrt(percent of engine power used) + constant)`
* `consumption = max_consumption * sqrt(percent of engine power used + constant)`
* `consumption = max_consumption * (percent of engine power used) / (gaussian with power usage as x axis + const)`

The constant factor should be big enough that it won't make much sense to start V12 to pull a cart. It should involve engine power, but could also involve engine type. Electric engines shouldn't penalize low speed as much as gasoline ones, for example.

In fact, the electric engines could always have perfect efficiency - this would make them useful for frequently varying speeds, like how they're used IRL in hybrid vehicles. One of the combustion engine types (gas or diesel) could have the opposite - increased penalty from being too far from optimum, to make hybrids more useful.

I'm thinking something like:
* At optimum, the efficiency is (obviously) 1.0
* At 0 and safe speed, the efficiency is some constant based on engine. Say, 0.2 for gas, 0.8 for electric
* Between those, symmetric gaussian, linear interpolation, rescaled tanh, `exp(-x)` or something like that

Though it shouldn't lead to situations where speeding up decreases fuel consumption per fuel consumption tick (only per distance traveled).

**Comment by mugling** [comment]
> In fact, the electric engines could always have perfect efficiency - this would make them useful for frequently varying speeds, like how they're used IRL in hybrid vehicles

That's an excellent idea and especially relevant after #15744

**Comment by mugling** [comment]
Could making `safe speed` dependent upon only the engine and displacement work?
IRL engines have a preferred RPM, the useful range of which is extended via gearing?

**Comment by mugling** [comment]
`1 - tanh( std::abs( ( veloticty - optimal ) ) / optimal )`

For optimal values of 60 (blue), 100 (red) and 140 (green) this is:

![ranh](https://cloud.githubusercontent.com/assets/12238606/13731659/bc077724-e966-11e5-86fb-e0e0f5b5ab68.png)

**Comment by Coolthulhu** [comment]
> Could making safe speed dependent upon only the engine and displacement work?

Sounds hard to balance. For example, V12 pulling a gigantic tank (and thus unable to achieve great speeds) vs. same V12 on a sports car.

> For optimal values of 60 (blue), 100 (red) and 140 (green) this is:

Looks a bit too strict around the optimum. That could make it hard to use, especially considering how currently velocity can be only specified in discrete increments.

Having the formula reach numbers higher than 1 and then capping it at 1 would make it easier to get the optimum. Also, driving skill could affect the values somehow.

**Comment by mugling** [comment]
> Looks a bit too strict around the optimum

Agreed, but how to adjust it?

> especially considering how currently velocity can be only specified in discrete increments

Is there any reason we have to maintain this?

> Also, driving skill could affect the values somehow.

Possibly but that can be a small effect we can add later

**Comment by Coolthulhu** [comment]
> Agreed, but how to adjust it?

The part about going above 1 and then capping could solve it. Alternatively, formulas like `exp(-abs(x)^3)` or `exp(-x^4)` seem to look good.

> Is there any reason we have to maintain this?

It's not a strict rule, as soft velocity cap around safe velocity ignores it, it's just that it's easier to have velocity in tiles/turn. Plus, turning looks really ugly at velocity lower than 16 km/h (10 mph).

**Comment by mugling** [comment]
For an optimal speed of 80km/h

![ranh2](https://cloud.githubusercontent.com/assets/12238606/13731962/30a6031e-e96e-11e5-9e6a-02255f0079b7.png)

**Comment by mugling** [comment]
When do we actually consume fuel?

https://gist.github.com/mugling/3ecd0e1c732273f27b72

**Comment by Coolthulhu** [comment]
That looks quite good.

How will this speed scaling interact with alternators?

> When do we actually consume fuel?

In `vehicle::thrust`, when accelerating. Currently the acceleration is in bursts rather than continuous.

**Comment by mugling** [comment]
> How will this speed scaling interact with alternators?

We could impose a minimum fuel usage for them. IRL you'd need a huge alternator to retard an engine with an open throttle. Generator sets of course would indeed have that huge minimum fuel usage to keep any attached engine occupied. Engines maintain an inbuilt idle.

**Comment by mugling** [comment]
Somehow we really need to go from `velocity` to `load` where the former is an absolute value and the latter a ratio.

How about scaling `load` from 0 to max_velocity() placing safe_velocity() at 50% and then interpolating from there?

And then using `load` as the input to the previous efficiency curve?

**Comment by mugling** [comment]
How does `velocity` compared to tiles traveled per move/turn?

**Comment by Coolthulhu** [comment]
> How does velocity compared to tiles traveled per move/turn?

I think 100 is tile per turn.

**Comment by mugling** [comment]
Is that the same as the displayed vehicle speed or is it rescaled?

**Comment by mugling** [comment]
What are some typical raw speeds for example vehicles?

**Comment by jcd000** [comment]
The final distribution and rationale sounds pretty sensible as is.

Also modeling acceleration on top of that could be simple: Assume the engine has to work harder than normal to accelerate, so use the efficiency for the speed selected, rather than the current speed.
This way, even if it may be working at a higher efficiency level, it will be draining per-turn as much as needed for the speed selected, even if the car will be moving slower than that until it reaches that speed.

**Comment by illi-kun** [comment]
Continued in #19348.
